django/db/models/manager.py

```diff
@@ -242,12 +244,30 @@ class Manager(BaseManager.from_queryset(QuerySet)):
 class ManagerDescriptor(object):
     # This class ensures managers aren't accessible via model instances.
     # For example, Poll.objects works, but poll_obj.objects raises AttributeError.
-    def __init__(self, manager):
+    # It also ensures managers defined on non-abstract base classes are not inherited by child classes.
+    def __init__(self, manager, model):
         self.manager = manager
+        self.model = model
 
     def __get__(self, instance, cls=None):
         if instance is not None:
             raise AttributeError("Manager isn't accessible via %s instances" % cls.__name__)
+
+        for model in cls.mro()[:-2]:
+            model_is_concrete = not (model._meta.abstract or model._meta.proxy or model._meta.swapped)
+            if model_is_concrete and cls._meta.concrete_model and model != cls._meta.concrete_model:
+                continue
+            manager_descriptor = model.__dict__.get(self.manager.name, None)
+            if isinstance(manager_descriptor, (ManagerDescriptor, AbstractManagerDescriptor)):
+                return manager_descriptor.manager
+
```

Comment:
@ar45, that's true and that's what it does now since `Model.objects` is a descriptor. For all cases but new class creation there's only `if cls._meta.concrete_model == self.model._meta.concrete_model:` addition in my PR. Do you think it should be cached? What kind of caching do you suggest then?